### PR TITLE
Call syncfs on the mount when fsync is called, powered by EMTERPRETIFY_ASYNC

### LINKED ISSUE #2492: Call syncfs() on IDBFS when the C fsync() is called
This seems reasonable to do, as IDBFS does have the a concept of writing out to storage. However, it is asynchronous which may be confusing.

### PR DESCRIPTION
**Pull request by timdream**
This should fix #2492.

I don't understand how I could setup a test case for this -- is there a test suite with `EMTERPRETIFY_ASYNC=1` that runs on the browser?

Please guide me on how to write a test for this.

### CODE REVIEW

**Comment by timdream** [comment]
I also wonder if I should modify the original function as well and make it call `syncfs` and don't wait for the callback (and the error code). It would be better than not saving the files at all.

(My original target was em-dosbox, which unfortunately doesn't work with EMTERPRETIFY_ASYNC=1 yet.)

File: src/library.js [hunk 1]
```diff
@@ -868,6 +868,35 @@ LibraryManager.library = {
...
+  fsync: function(fildes) {
+    return EmterpreterAsync.handle(function(resume) {
+      // int fsync(int fildes);
+      // http://pubs.opengroup.org/onlinepubs/000095399/functions/fsync.html
+      var stream = FS.getStream(fildes);
+      if (stream) {
+        var mount = stream.node.mount;
+        if (!mount.type.syncfs) {
+          // We write directly to the file system, so there's nothing to do here.
+          resume(function() { return 0 });
```
Comment (coolwanglu) [inline src/library.js]: I don't think you can directly call `resume` in `handle`, which should trigger an error of bad state of EmterpreterAsync.
The best solution I have is to call it with `setTimeout`

Comment (kripken) [inline src/library.js]: Yes, the current code assumes `handle` does actually trigger an async operation. But we could make resume force it to be async, so you don't need to handle that here, I'll do that now.

**Comment by kripken** [comment]
Looks nice!

For a test, see `test_emterpreter_async` in `tests/test_browser.py`. The second argument to `btest` is the output from `REPORT_RESULT` in the test.

Please test both synchronous and asynchronous calls to `resume`.

File: src/library.js (hunk 1 above)
Comment (kripken) [inline src/library.js]: I pushed that to incoming.

**Comment by timdream** [comment]
@kripken Thank you very much for your comment and help. I am updating my progress so far:

1. It turned out `fclose` also calls `_fsync` directly in `library.js`, so I would have to wrap that function too.  emterpreter-async says it can't save the stack with that. I am pushing a new commit on that just a moment.
2. As of testing, my plan right now is to copy and edit `test_idbfs_sync.c` and create one for `fclose` and one for `fsync` to test the actual feature (and async calls to `resume`). However I hit the same issue when running the test, apparently because of how the tests are executed:

```
uncaught exception: abort("Assertion failed: cannot start async op with normal JS calling ccall") at jsStackTrace@http://localhost:8888/test.js:1158:13
stackTrace@http://localhost:8888/test.js:1175:22
abort@http://localhost:8888/test.js:12285:44
assert@http://localhost:8888/test.js:357:5
ccallFunc@http://localhost:8888/test.js:425:1
Runtime.asmConsts</<@http://localhost:8888/test.js:1533:139
done@http://localhost:8888/test.js:3077:13
IDBFS.reconcile@http://localhost:8888/test.js:2415:1
IDBFS.syncfs/</<@http://localhost:8888/test.js:2215:13
@http://localhost:8888/test.js:2309:1

This error happened during an emterpreter-async save or load of the stack. Was there non-emterpreted code on the stack during save (which is unallowed)? You may want to adjust EMTERPRETIFY_BLACKLIST, EMTERPRETIFY_WHITELIST, or EMTERPRETIFY_YIELDLIST (to consider certain functions ok to run during an emscripten_sleep_with_yield).
This is what the stack looked like when we tried to save it: 1,EmterpreterAsync.handle@http://localhost:8888/test.js:5344:40
_fsync@http://localhost:8888/test.js:5369:1
emterpret@http://localhost:8888/test.js:9982:11
_test@http://localhost:8888/test.js:10303:2
asm._test@http://localhost:8888/test.js:10392:8
ccallFunc@http://localhost:8888/test.js:423:15
Runtime.asmConsts</<@http://localhost:8888/test.js:1533:139
done@http://localhost:8888/test.js:3077:13
IDBFS.reconcile@http://localhost:8888/test.js:2415:1
IDBFS.syncfs/</<@http://localhost:8888/test.js:2215:13
@http://localhost:8888/test.js:2309:1
```

I will see what can be done about this.
3. As of sync call to `resume`, I think I can either create MEMFS fsync/fclose tests or a specific test case for it.

**Comment by timdream** [comment]
I am now been able to run the tests by moving initial `FS.syncfs` to `pre.js`.

**Comment by timdream** [comment]
I have completed all the tests intended and verify that they passes on the latest Nightly locally. Please tell me what I could improve before merge.

I also don't know if this project require squashing commits. If so I will do so.

File: tests/fs/test_idbfs_fclose.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    if (close(fd) != 0)
+      result = -5000 - errno;
+    if (unlink("/working1/wakakaka.txt") != 0)
+      result = -6000 - errno;
+  }
+
+  // sync from memory state to persisted and then
+  // run 'success'
+  EM_ASM(
+    FS.syncfs(function (err) {
```
Comment (timdream) [inline tests/fs/test_idbfs_fclose.c]: This is needed because `unlink` does not persist on its own.

File: src/library.js [hunk 3]
```diff
@@ -1936,6 +1966,33 @@ LibraryManager.library = {
...
+      // and only returns until syncfs completes.
+      stream = FS.getStreamFromPtr(stream);
+      if (stream) {
+        var mount = stream.node.mount;
+        if (!mount.type.syncfs) {
+          resume(function() { _close(fd); });
+          return;
+        }
+        mount.type.syncfs(mount, false, function(err) {
+          // disregard err
```
Comment (timdream) [inline src/library.js]: According to [the spec](http://pubs.opengroup.org/onlinepubs/000095399/functions/fclose.html) it's better to handle the error, if you think so I will do it.

**Comment by kripken** [comment]
I am concerned about the duplication of code in the two asyncified methods - the internal logic that is the same in the two cases is now in 2 places. It is rather small though, so maybe I am fretting about nothing.

This makes fclose call fsync - but is that correct? I think people need to call fsync if they want it, I don't remember the spec saying fclose will fsync for you - or am I wrong?

**Comment by timdream** [comment]
> I am concerned about the duplication of code in the two asyncified methods - the internal logic that is the same in the two cases is now in 2 places. It is rather small though, so maybe I am fretting about nothing. (We can call handle() conditionally, which might allow more code sharing.) Thoughts?

Without changing how `handle()` would work, I could concentrate logic into a `__fsyncAsync` method instead. Would that be better in your opinion?

> This makes fclose call fsync - but is that correct? I think people need to call fsync if they want it, I don't remember the spec saying fclose will fsync for you - or am I wrong?

The original `fclose` function calls `fsync`, also the spec said "The fclose() function shall cause the stream pointed to by stream to be flushed and the associated file to be closed." I don't know by *flush* it means something equivalent to `fsync`. If it's not the call should be removed from the original function as well.

File: src/library.js [hunk 4]
```diff
@@ -1936,6 +1972,25 @@ LibraryManager.library = {
...
+  fclose: function(stream) {
+    return EmterpreterAsync.handle(function(resume) {
+      // int fclose(FILE *stream);
+      // http://pubs.opengroup.org/onlinepubs/000095399/functions/fclose.html
+      var fd = _fileno(stream);
+
+      ___fsyncAsync(fd, function(errno) {
+        if (errno) {
+          resume(function() { return errno; });
+          return;
```
Comment (timdream) [inline src/library.js]: I ended up not calling `_close` in this early return block.

**Comment by timdream** [comment]
Just pushed a commit (tests passes) for that `__fsyncAsync` idea.

**Comment by kripken** [comment]
Yes, flush is not the same as fsync. fsync is guaranteed to write to external storage, flush just pushes currently queued stuff in libc, essentially. I think it caused no harm for us to call fsync there because it wasn't really doing anything, it was a no-op. But if fsync is now going to be fixed, we should not call it from fclose.

**Comment by kripken** [comment]
Given that, I think the last commit isn't needed.

**Comment by timdream** [comment]
Understood. I've moved the original commits into a `fsync-emterpretify-async2` branch (as backup) and force pushed this branch to do what's needed accordingly.

File: tests/fs/test_memfs_fsync.c [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+#include <sys/stat.h>
+#include <errno.h>
+#include <string.h>
+
+int result = 1;
+
+int main() {
+  int fd;
+
+  // We first make sure the file doesn't currently exist
```
Comment (kripken) [inline tests/fs/test_memfs_fsync.c]: this comment looks wrong?

File: tests/fs/test_idbfs_fsync.c [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      result = -8000;
+    if (close(fd) != 0)
+      result = -9000 - errno;
+    if (unlink("/working1/wakaka.txt") != 0)
+      result = -10000 - errno;
+  }
+
+  // sync from memory state to persisted and then
+  // run 'success'
+  EM_ASM(
```
Comment (kripken) [inline tests/fs/test_idbfs_fsync.c]: just to be sure, this test uses a C fall to fsync, and a JS call to syncfs - is that intentional, to test both paths? I was expecting to see C for both.

File: tests/fs/test_memfs_fsync.c (hunk 5 above)
Comment (timdream) [inline tests/fs/test_memfs_fsync.c]: This comment is referring to `stat()` call in which I copied from `test_idbfs_syncfs.c`.

File: tests/fs/test_idbfs_fsync.c (hunk 6 above)
Comment (timdream) [inline tests/fs/test_idbfs_fsync.c]: This part runs at the end of `!FIRST` test run to ensure we also `unlink()` the file in IndexedDB.

**Comment by timdream** [comment]
Another question: The current patch does not `syncfs` for applications that operates on file streams. Should we find an opportunity in these operations that could trigger `syncfs`? The spec doesn't say `fclose()` implies `fsync` but it does not clearly say when should the stream op to be write on the disk either...

I can create another pull request for that if you could tell me where to call.

File: tests/fs/test_memfs_fsync.c (hunk 5 above)
Comment (kripken) [inline tests/fs/test_memfs_fsync.c]: hmm, but there is no `FIRST` in this entire file?

File: tests/fs/test_idbfs_fsync.c (hunk 6 above)
Comment (kripken) [inline tests/fs/test_idbfs_fsync.c]: ok

**Comment by kripken** [comment]
I don't think anything will implicitly fsync. fsync is a costly operation, so people writing native code must write it explicitly.

File: tests/fs/test_memfs_fsync.c (hunk 5 above)
Comment (timdream) [inline tests/fs/test_memfs_fsync.c]: Ah, make sense.

Comment (timdream) [inline tests/fs/test_memfs_fsync.c]: Fixed.

**Comment by kripken** [comment]
Looks good, please just add yourself to AUTHORS before we merge.

**Comment by timdream** [comment]
Done.

**Comment by kripken** [comment]
Merged, thanks!

**Pull request merged**